### PR TITLE
feat: define the turn for submitting attestation

### PR DESCRIPTION
**Pull request by forcodedancing**
### Description

This pr will define the turn for submitting challenge attestations.

### Rationale

To avoid offline challengers submit attestation at the same time, to avoid unnecessary gas cost.

### Example

NA

### Changes

Notable changes: 
* only challenger address can submit attestations

### CODE REVIEW

File: x/challenge/keeper/bls_signed_msg.go [hunk 1]
```diff
@@ -24,13 +25,7 @@ type BlsSignedMsg interface {
...
 
 // verifySignature verifies whether the signature is valid or not.
-func (k Keeper) verifySignature(ctx sdk.Context, signedMsg BlsSignedMsg) ([]string, error) {
-	historicalInfo, ok := k.stakingKeeper.GetHistoricalInfo(ctx, ctx.BlockHeight())
-	if !ok {
-		return nil, errors.Wrap(types.ErrInvalidVoteValidatorSet, "fail to get validators")
-	}
-	validators := historicalInfo.Valset
-
+func (k Keeper) verifySignature(ctx sdk.Context, signedMsg BlsSignedMsg, validators []stakingtypes.Validator) ([]string, error) {
```
[the remark this answers is not in the document]
Comment (forcodedancing) [reply x/challenge/keeper/bls_signed_msg.go]: removed

File: x/challenge/keeper/challenge.go [hunk 2]
```diff
@@ -69,24 +69,45 @@ func getChallengeKeyBytes(challengeId uint64) []byte {
...
 	store := prefix.NewStore(ctx.KVStore(k.storeKey), []byte{})
 	bz := store.Get(types.AttestChallengeIdKey)
 
 	if bz == nil {
-		return 0
+		return []uint64{}
 	}
 
-	return binary.BigEndian.Uint64(bz)
+	ids := []uint64{}
```
[the remark this answers is not in the document]
Comment (forcodedancing) [reply x/challenge/keeper/challenge.go]: > the static check suggest that change to `var ids []uint64`

Fixed. By the way, which tool you are using for this?

File: x/challenge/keeper/challenge.go [hunk 3]
```diff
@@ -69,24 +69,45 @@ func getChallengeKeyBytes(challengeId uint64) []byte {
...
-	bz := make([]byte, 8)
-	binary.BigEndian.PutUint64(bz, challengeId)
-	store.Set(types.AttestChallengeIdKey, bz)
+	bz := store.Get(types.AttestChallengeIdKey)
+
+	idBz := make([]byte, 8)
+	binary.BigEndian.PutUint64(idBz, challengeId)
+
+	bzLen := uint64(len(bz))
+	keepLen := maxToKeep * 8
```
[the remark this answers is not in the document]
Comment (forcodedancing) [reply x/challenge/keeper/challenge.go]: updated the default value

**Comment by forcodedancing** [comment]
> Please try to optimize the data structure to save storage

try to use a circular queue to store the latest attested challenge ids

**Review by guagualvcha — APPROVED** [review]

**Pull request merged**